### PR TITLE
Add a test for master restart after being killed

### PR DESCRIPTION
**Pull request by pkubatrh**
This PR adds a simple test to check if a master can be properly restarted after being killed in a replication scenario.
First commit makes some necessary changes to existing test code while the second commit adds the actual test.

### CODE REVIEW

**Comment by pkubatrh** [comment]
[test]

**Comment by bparees** [comment]
[test]

**Comment by praiskup** [comment]
Thanks.  I wouldn't use a word "failover" as that usually means different usecase in PostgreSQL context.  I'll do a better review tomorrow.

File: hack/run_test [hunk 1]
```diff
@@ -27,6 +24,13 @@ function cleanup() {
     docker rm $CONTAINER
     rm $cidfile
     echo "Done."
+}
+
+function cleanup() {
+  for cidfile in $CIDFILE_DIR/* ; do
+    CONTAINER=$(cat $cidfile)
+
+    cleanup_container
```
Comment (praiskup) [inline hack/run_test]: I would move this into the second patch, otherwise this "function split" looks like
unnecessary (when somebody looks at the patch separately).

File: hack/run_test [hunk 2]
```diff
@@ -259,7 +263,7 @@ function run_tests() {
 function run_slave() {
   local master_ip="$1"; shift
   local suffix="$1"; shift
-  docker run $cluster_args -e POSTGRESQL_MASTER_IP=${master_ip} \
+  docker run $cluster_args -p 5432 -e POSTGRESQL_MASTER_IP=${master_ip} \
```
Comment (praiskup) [inline hack/run_test]: Why is this needed? (missing comment probably).

Can we connect the master/slave containers without exposing ports from host system?

File: hack/run_test [hunk 3]
```diff
@@ -311,6 +311,52 @@ function setup_replication_cluster() {
...
 
+function run_failover_test {
+  echo "Testing failover"
+  local cluster_args="-e POSTGRESQL_ADMIN_PASSWORD=pass -e POSTGRESQL_MASTER_USER=$PGUSER -e POSTGRESQL_MASTER_PASSWORD=$PASS"
+  local cid_suffix="failover"
+
+  # Setup the cluster
+  setup_replication_cluster
+
+  # Kill the master
```
Comment (praiskup) [inline hack/run_test]: Could we use something like `: Kill the master`, instead of `# Kill the master`...  so we can see from log what is happening during 'make test'.

File: hack/run_test (hunk 2 above)
Comment (pkubatrh) [inline hack/run_test]: Thanks for spotting this. This is actually leftover code from me trying to figure out why the the new slave doesn't connect to the master. 
Will remove the change in the next push as it is not needed.

File: hack/run_test (hunk 3 above)
Comment (praiskup) [inline hack/run_test]: Also, I would consider `docker kill` to really shoot the postgres server, `docker stop` rather carefully stops the server.

**Comment by praiskup** [comment]
Ah, reading the patch carefully, there really is failover scenario tested (that is, however, probably not usable in OpenShift at least ATM).

**Comment by praiskup** [comment]
IOW: promoting slave container to master container is hard task enough to rather start new
master...

**Comment by pkubatrh** [comment]
Updated the PR with a new test case (master restart instead of failover).

File: hack/run_test [hunk 4]
```diff
@@ -308,21 +294,28 @@ function run_replication_test() {
...
+
+function test_value_replication() {
+  local table_name="$1"; shift
 
+  # Setup the replication data
   local value
   value=24
-  # Do some real work to test replication in practice
-  postgresql_cmd -c "CREATE TABLE t1 (a integer); INSERT INTO t1 VALUES ($value);"
+  local table_name="t1"
```
Comment (praiskup) [inline hack/run_test]: This overwrites previous variable assignment.

File: hack/run_test [hunk 5]
```diff
@@ -338,6 +331,37 @@ function run_replication_test() {
...
+  max_attempts=30
+
+  # Setup the cluster
+  setup_replication_cluster
+
+  # Do some real work to test replication in practice
+  test_value_replication $table_name
+}
+
```
Comment (praiskup) [inline hack/run_test]: Please don't move the functions, if possible.  It is harder to review (from the diff is not obvious what changed within function).

File: hack/run_test [hunk 6]
```diff
@@ -308,21 +294,28 @@ function run_replication_test() {
...
   local slave_ip
+  local cidfile
+  local master_ip=$(get_container_ip master-$cid_suffix.cid)
   for f in `find ${CIDFILE_DIR} -type f -name "slave-*"`; do
     cidfile=$(basename $f)
     slave_ip=$(get_container_ip $cidfile)
     CONTAINER_IP=$slave_ip
     for i in $(seq $max_attempts); do
-      result="$(postgresql_cmd -At -c "select * from t1")"
+      result="$(postgresql_cmd -At -c "select * from $table_name")"
```
Comment (praiskup) [inline hack/run_test]: Once there are variables, you can consider 'psql --set' option.  We are in tests only, so it is safe so far.

File: hack/run_test [hunk 7]
```diff
@@ -338,6 +331,37 @@ function run_replication_test() {
...
+  CONTAINER_IP=$master_ip
+  test_slave_visibility $cid_suffix
+}
+
+function run_replication_test() {
+  echo "Testing master-slave replication"
+  local cluster_args="-e POSTGRESQL_ADMIN_PASSWORD=pass -e POSTGRESQL_MASTER_USER=$PGUSER -e POSTGRESQL_MASTER_PASSWORD=$PASS"
+  local cid_suffix="basic"
+  local table_name="t1"
+  max_attempts=30
```
Comment (praiskup) [inline hack/run_test]: Not a local variable, it is not obvious this is expected.

**Comment by pkubatrh** [comment]
New push, fixed issues pointed out by @praiskup (overwritten variable + added a comment to global variable)

File: hack/run_test [hunk 8]
```diff
@@ -308,21 +294,28 @@ function run_replication_test() {
...
+  local cidfile
+  local master_ip=$(get_container_ip master-$cid_suffix.cid)
   for f in `find ${CIDFILE_DIR} -type f -name "slave-*"`; do
     cidfile=$(basename $f)
     slave_ip=$(get_container_ip $cidfile)
     CONTAINER_IP=$slave_ip
     for i in $(seq $max_attempts); do
-      result="$(postgresql_cmd -At -c "select * from t1")"
+      result="$(postgresql_cmd -At -c "select * from $table_name")"
       if [[ ! -z "${result}" ]]; then
```
Comment (praiskup) [inline hack/run_test]: This is probably OK, but I would rather fix it sooner than later ... checking psql output 
for (non)empty-ness is not perfect, because error messages comes to standard output ->
thus empty string comes from invalid query.  We are fine because 'set -e' is activated ..
but that could change in future...  I would vote for explicit string to compare $result with.

File: hack/run_test [hunk 9]
```diff
@@ -348,11 +350,51 @@ function setup_replication_cluster() {
...
+  docker run $master_args --rm $IMAGE_NAME /bin/sh -c "$rm_cmd"
+  rmdir ${volume_dir}
+}
+
 function run_replication_test() {
   echo "Testing master-slave replication"
   local cluster_args="-e POSTGRESQL_ADMIN_PASSWORD=pass -e POSTGRESQL_MASTER_USER=$PGUSER -e POSTGRESQL_MASTER_PASSWORD=$PASS"
   local cid_suffix="basic"
   local table_name="t1"
+  # Global variable for setup_replication_cluster and test_value_replication
```
Comment (praiskup) [inline hack/run_test]: Why you can't use local variable?

Comment (pkubatrh) [inline hack/run_test]: With a setting such as this that is used in polling loops and probably won't be changed much from its default value, I felt like it would be best to use it as a global variable. Although with the addition of another test I am now setting it in two places so I will fix that.

Comment (pkubatrh) [inline hack/run_test]: Scratch that, I see it's found in other parts of the code as well. Will make it local then.

**Comment by pkubatrh** [comment]
Rewrote non-emptyness check into an expected value check, max_attempts variable no longer global.

File: hack/run_test [hunk 10]
```diff
@@ -350,6 +352,43 @@ function setup_replication_cluster() {
...
+  test_slave_visibility $cid_suffix
+
+  # Check if the replication works
+  test_value_replication $table_name
+
+  # Cleanup the volume
+  docker stop $(cat $cidfile)
+  local rm_cmd='/bin/rm -rf /var/lib/pgsql/data/*'
+  docker run $master_args --rm $IMAGE_NAME /bin/sh -c "$rm_cmd"
+  rmdir ${volume_dir}
```
Comment (praiskup) [inline hack/run_test]: Nit, can we do this at one place (looks like code duplication).

File: hack/run_test [hunk 11]
```diff
@@ -350,6 +352,43 @@ function setup_replication_cluster() {
...
+  local table_name="t1"
+
+  local volume_dir
+  volume_dir=`mktemp -d --tmpdir pg-testdata.XXXXX`
+  chmod a+rwx ${volume_dir}
+  master_args="-v ${volume_dir}:/var/lib/pgsql/data:Z"
+
+  # Setup the cluster
+  local slave_num=2
+  setup_replication_cluster $slave_num
```
Comment (praiskup) [inline hack/run_test]: Looks like setup_replication_cluster takes the $slave_num twice, once as variable, once as argument.

File: hack/run_test [hunk 12]
```diff
@@ -265,7 +265,8 @@ function run_slave() {
 
 function run_master() {
   local suffix="$1"; shift
-  docker run $cluster_args -p 5432  \
+  master_args=${master_args-}
+  docker run $cluster_args $master_args -p 5432  \
```
Comment (praiskup) [inline hack/run_test]: I'm still not sure whether we need to '-p 5432'.

File: hack/run_test [hunk 13]
```diff
@@ -308,21 +295,29 @@ function run_replication_test() {
...
-  postgresql_cmd -c "CREATE TABLE t1 (a integer); INSERT INTO t1 VALUES ($value);"
+  local table_name="t1"
+  postgresql_cmd -c "CREATE TABLE $table_name (a integer); INSERT INTO $table_name VALUES ($value);"
 
   # Read value from slaves and check whether it is expected
   local f
   local slave_ip
+  local cidfile
+  local master_ip=$(get_container_ip master-$cid_suffix.cid)
   for f in `find ${CIDFILE_DIR} -type f -name "slave-*"`; do
```
Comment (praiskup) [inline hack/run_test]: Here is probably missing "suffix" in find command.  Btw., I would much rather see the `run_slave` method "appending" the IP of newly created slave into let's say $slave_addresses variable.  Then you can simply iterate like `for i slave in $slave_addresses`.

Comment (pkubatrh) [inline hack/run_test]: Good idea. It would be better to take this one step further though and append the container's ID instead. We only need the IPs now, but we might need something else in the future.
We will still need unique file names for the cid files in order for the containers to be cleaned up properly, so the $suffix will probably need to stay in.

File: hack/run_test (hunk 11 above)
Comment (pkubatrh) [inline hack/run_test]: setup_replication_cluster takes the local $slave_num variable as an optional argument and sets up a local variable with the same name inside the function, defaulting to 1 if called without the argument.
As both of the variables are local, there should not be any value transfer by variable going on.

Comment (praiskup) [inline hack/run_test]: It then just looks weird (or complicated), because you are overwriting the `slave_num` from `run_master_restart_test` with another local variable (named exactly the same) ...  But the local variable from `run_master_...` is available in called `setup_replication_cluster` function, too.   Also, the local variable is not needed:
https://github.com/praiskup/openshift-postgresql/commit/c1b99705a2

File: hack/run_test (hunk 13 above)
Comment (praiskup) [inline hack/run_test]: > We only need the IPs now, but we might need something else in the future.

The point was not to docker inspect all the time, but that is just preference.  Container's id is better than doing error-prone find command.

> We will still need unique file names for the cid files in order for the containers to be cleaned up properly, so the $suffix will probably need to stay in.

Right, the suffix is already there.  But this particular find could go away completely.

**Comment by pkubatrh** [comment]
Incorporated @praiskup's suggestions - new functions for the creation and removal of volumes, got rid of some redundant local variables, changed loops to use container IDs instead of calling 'find' on .cid files.

**Comment by praiskup** [comment]
Weird.  On my Fedora 24 box, running 'make test VERSION=9.4' fails very early with:
+ timeout -s 9 --preserve-status 60s docker run --rm -e POSTGRESQL_USER=user -e POSTGRESQL_PASSWORD=pass centos/postgresql-94-centos7-candidate
/bin/bash: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: Permission denied

Actually, the docker build created unusable image, the source 'centos:centos7' works fine.

**Comment by praiskup** [comment]
[test]

**Comment by praiskup** [comment]
My mistake, more info in #136, I'll continue with the review.

File: hack/run_test [hunk 14]
```diff
@@ -309,7 +328,7 @@ function test_value_replication() {
     CONTAINER_IP=$slave_ip
     for i in $(seq $max_attempts); do
       result="$(postgresql_cmd -At -c "select * from $table_name")"
```
Comment (praiskup) [inline hack/run_test]: I'm not sure the loop here has some effect.  When the value is not there yet, it is
very likely the $table_name table does not exist yet too.  Then, postgresql_cmd fails
and this would fail the whole command, so no other attempt would me made.

Probably `|| :` is missing.

Comment (praiskup) [inline hack/run_test]: s/fail the whole command/exit the whole script/

**Comment by pkubatrh** [comment]
Pushed new changes - ignore psql errors when checking for value replication, master_args variable now local

**Comment by praiskup** [comment]
LGTM.

**Comment by praiskup** [comment]
@pkubatrh looks like this is pretty atomic change, would it be fine to squash the commits?

**Comment by bparees** [comment]
squash merged.
